### ISSUE TITLE
#[may_dangle] compiles on non-Drop impls

### ISSUE THREAD

**Issue by Ixrec** [rust-lang/rust#68411]
[The RFC](https://github.com/rust-lang/rfcs/pull/1327) pretty strongly implies that this attribute is only supposed to be used on Drop, and the semantics of it don't make any sense on other traits. Yet, this compiles:
```rs
#![feature(dropck_eyepatch)]

struct Foo<T>(T);

unsafe impl<#[may_dangle] T> std::ops::Deref for Foo<T> {
    type Target = T;
    fn deref(&self) -> &T {
        &self.0
    }
}
```
[playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d6bd53017167935e14968b885bf14598)

**Comment by kei519** [comment]
This seems to be fixed with [3a9bf4551374893fdc522572ee569028186e22cc](https://github.com/rust-lang/rust/commit/3a9bf4551374893fdc522572ee569028186e22cc)?

**Comment by jieyouxu** [comment]
Fixed by https://github.com/rust-lang/rust/pull/129235
